### PR TITLE
Don't crash into floor

### PR DESCRIPTION
**Pull request by Finn6360**
Makes it not crash into floor.

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/Arm.java [hunk 1]
```diff
@@ -7,19 +7,32 @@
...
 
     public Data data = new Data();
-    public ArmMap map;
+    public ArmMap extendMap;
     public final double SPEED = 0.3;
     private final double RETRACT_SPEED = -0.1;
+    final double PIVIOT_HEIGHT = 46.654;
+    private double ARM_ANGLE;
+
+    NetworkTableInstance inst = NetworkTableInstance.create();
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Arm.java]: Can this be done on startup? Or does it have to wait for something else to come up? Also you're creating an entire table?

File: src/main/java/frc/robot/subsystems/Arm.java [hunk 2]
```diff
@@ -7,19 +7,32 @@
...
+    NetworkTable table = inst.getTable("ArmTable");
+
+    public Arm(ArmMap extendMap) {
+        this.extendMap = extendMap;
+    }
+
+    public boolean intakeBelowGround() {
+        return (Math.cos(ARM_ANGLE) * data.distanceInches) - PIVIOT_HEIGHT < 0;
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Arm.java]: `x - y < 0` is just `x < y` for people who want to look fancy

File: src/main/java/frc/robot/subsystems/Arm.java [hunk 3]
```diff
@@ -7,19 +7,32 @@
...
 
 public class Arm extends SmartSubsystemBase {
 
     public Data data = new Data();
-    public ArmMap map;
+    public ArmMap extendMap;
     public final double SPEED = 0.3;
     private final double RETRACT_SPEED = -0.1;
+    final double PIVIOT_HEIGHT = 46.654;
+    private double ARM_ANGLE;
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Arm.java]: Capitals are for constants. And cookies are for closers.

File: src/main/java/frc/robot/subsystems/Arm.java [hunk 4]
```diff
@@ -113,14 +126,19 @@ public void safeState() {
...
+        ARM_ANGLE = table.getValue("angle").getDouble();
     }
 
     // Adds softlimit to arm extension speed
     private double softLimit(double speed) {
-        if ((data.distanceInches > map.softMaxDistance && speed > 0) || (data.distanceInches < map.softMinDistance
-                && speed < 0)) {
+        if ((data.distanceInches > extendMap.softMaxDistance && speed > 0)
+                || (data.distanceInches < extendMap.softMinDistance && speed < 0)) {
+            if (speed > 0 && intakeBelowGround()) {
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Arm.java]: This if statement could be before the previous block

File: src/main/java/frc/robot/subsystems/ArmRotate.java [hunk 5]
```diff
@@ -61,11 +74,17 @@ public void periodic() {
...
         Logger.getInstance().processInputs(getName(), data);
+        table.putValue("angle", NetworkTableValue.makeDouble(data.degrees));
+        ARM_LENGTH = table.getValue("armLegnth").getDouble();
     }
 
     private double softLimit(double speed) {
         if ((data.degrees > this.map.topAngle && speed > 0) ||
                 (data.degrees < this.map.bottomAngle && speed < 0)) {
+
+            if (speed < 0 && intakeBelowGround()) {
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/ArmRotate.java]: This if should be before the enclosing if

File: src/main/java/frc/robot/subsystems/ArmRotate.java [hunk 6]
```diff
@@ -32,6 +40,11 @@ public CommandBase move(DoubleSupplier rotationSpeed) {
         });
     }
 
+    public boolean intakeBelowGround() {
+        return (Math.cos(data.degrees) * ARM_LENGTH) - PIVIOT_HEIGHT < 0;
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/ArmRotate.java]: `x - y < 0`

File: src/main/java/frc/robot/subsystems/ArmRotate.java [hunk 7]
```diff
@@ -17,8 +20,13 @@ public class ArmRotate extends SmartSubsystemBase {
     final double MOVE_SPEED = 0.5;
     final double COMPARE_ANGLE = 5;
     final double SLOW_DOWN = 0.1;
+    final double PIVIOT_HEIGHT = 46.654;
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/ArmRotate.java]: This is not how "pivot" is spelled

**Review by msoucy — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Arm.java (hunk 1 above)
Comment (msoucy) [reply src/main/java/frc/robot/subsystems/Arm.java]: I *highly* recommend using `getDefault` here, per [the documentation](https://docs.wpilib.org/en/stable/docs/software/networktables/tables-and-topics.html)

**Comment by msoucy** [comment]
Since the encoders get zeroed on startup, the safest thing to do would be to have the encoder get reset when the arm is fully extended or retracted.
When the current in the motor is way too high and the speed is negative, then it's probably trying to retract but it's all the way retracted. In that case, set the encoder value to 0.
When the current in the motor is way too high and the speed is positive, then it's probably trying to extend but it's all the way extended. In that case, set the encoder to `MAX_EXTENDED` value.

**Comment by TimB-87** [comment]
I believe I did what you wanted Matt. We had the retract stopper and Ben said we probably don't need the extend stopper. I did, however, add hard limits.

**Pull request closed without merging**